core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -630,128 +673,115 @@ void freeze() {
-                    &&
-                    (sortedKeyIndex + keyLengthToCopy <= maxBytes)
-                    &&
-                    (prevKeyIndex + prevKeyLength == currKeyIndex))) {
-
-                entryIndexEnd++;
-                isFirst = false;
-                eiPrev = ei;
-                ei = srcChunk.get(ei, OFFSET_NEXT);
-                prevKeyIndex = currKeyIndex;
-                prevKeyLength = srcChunk.get(eiPrev, OFFSET_KEY_LENGTH);
-                keyLengthToCopy += prevKeyLength;
-                if (ei != NONE) continue;
+            // we cannot enlarge interval: if key is removed (handle index is -1) or
+            // if this chunk already has all entries to start with
+            if ((currSrcHandleIndex > 0) && (sortedEntryIndex + entriesToCopy * FIELDS <= maxIdx)) {
+                // we can enlarge the interval, if it is otherwise possible:
+                // if this is first entry in the interval (we need to copy one entry anyway) OR
+                // if (on the source chunk) current entry idx directly follows the previous entry idx
+                if (isFirstInInterval || (srcPrevEntryIdx + FIELDS == srcEntryIdx)) {
+                    entryIndexEnd++;
+                    isFirstInInterval = false;
+                    srcPrevEntryIdx = srcEntryIdx;
+                    srcEntryIdx = srcChunk.getEntryField(srcEntryIdx, OFFSET_NEXT);
+                    currSrcHandleIndex = srcChunk.getEntryField(srcEntryIdx, OFFSET_HANDLE_INDEX);
+                    if (srcEntryIdx != NONE) continue;
+
+                }
             }
 
-            itemsToCopy = entryIndexEnd - entryIndexStart + 1;
-            if (itemsToCopy > 0) { // copy continuous interval (with arraycopy)
-                // first copy key array
-                int sortedKI = sortedKeyIndex;
-                for (int i = 0; i < itemsToCopy; ++i) {
+            entriesToCopy = entryIndexEnd - entryIndexStart + 1;
+            if (entriesToCopy > 0) { // copy continuous interval (TODO: with arraycopy)
+                for (int i = 0; i < entriesToCopy; ++i) {
                     int offset = i * FIELDS;
                     // next should point to the next item
-                    entries[sortedEntryIndex + offset + OFFSET_NEXT] = sortedEntryIndex + offset + FIELDS;
-                    entries[sortedEntryIndex + offset + OFFSET_KEY_INDEX] = sortedKI;
-                    int keyLength = srcChunk.entries[entryIndexStart + offset + OFFSET_KEY_LENGTH];
-                    entries[sortedEntryIndex + offset + OFFSET_KEY_LENGTH] = keyLength;
-                    sortedKI += keyLength;
+                    entries[sortedEntryIndex + offset + OFFSET_NEXT]
+                        = sortedEntryIndex + offset + FIELDS;
+
+                    // copy the values of key position, key block index + key length => 2 integers via array copy
+                    System.arraycopy(srcChunk.entries,  // source array
```

Comment:
is this just copying a single entry via arraycopy?

Reply:
Yes, it is a copy of 2 integers in one assignment versus 2 separate assignments of each integer. Once we take Handles to off-heap it can be 3 integers. Not a big improvement, I agree.